cde/programs/dtksh/ksh93/src/lib/libast/man/fmtls.3
.\" $XConsortium: fmtls.3 /main/2 1996/10/29 15:07:55 drk $
.de L		\" literal font
.ft 5
.it 1 }N
.if !\\$1 \&\\$1 \\$2 \\$3 \\$4 \\$5 \\$6
..
.de LR
.}S 5 1 \& "\\$1" "\\$2" "\\$3" "\\$4" "\\$5" "\\$6"
..
.de RL
.}S 1 5 \& "\\$1" "\\$2" "\\$3" "\\$4" "\\$5" "\\$6"
..
.de EX		\" start example
.ta 1i 2i 3i 4i 5i 6i
.PP
.RS 
.PD 0
.ft 5
.nf
..
.de EE		\" end example
.fi
.ft
.PD
.RE
.PP
..
.TH STRLS 3
.SH NAME
fmtls \- format file information in buffer
.SH SYNOPSIS
.EX
#include <ls.h>

char*         fmtls(char* \fIbuf\fP, char* \fIname\fP, struct stat* \fIst\fP,
                    char* \fIinfo\fP, char* \fIlink\fP, int \fIflags\fP);
.EE
.SH DESCRIPTION
.L fmtls
formats
.IR ls (1)
style file information into the buffer
.IR buf .
A pointer to the trailing 0 in 
.I buf
is returned.
.I name
is the file name.
.I st
is a pointer to the
.B "struct stat
status information for
.I name
obtained from one of the
.IR stat (2)
routines.
.I info
is an additional string that will appear before
.I name
in
.I buf
and
.I link
is the name of the related hard or soft link file.
Both
.I info
and
.I link
may be 0.
.I flags
controls the format style and may be a combination of the following:
.PP
.TP
.B LS_ATIME
Use
.I st->st_atime
rather than
.I st->st_mtime
for
.BR LS_LONG .
.TP
.B LS_CTIME
Use
.I st->st_mtime
rather than
.I st->st_mtime
for
.BR LS_LONG .
.TP
.B LS_BLOCKS
List the number of blocks.
.TP
.B LS_INUMBER
List the file serial number (inode number).
.TP
.B LS_LONG
List the file mode, link count, user and group name,
size or major/minor device number, and date along with the
file name.
.TP
.B LS_MARK
The file name is appended with
.L /
for directories,
.L @
for symbolic links,
and
.L *
for executable files.
.TP
.B LS_NOGROUP
Omit the group name from
.BR LS_LONG .
.TP
.B LS_NOUSER
Omit the user name from
.BR LS_LONG .
.PP
The user and group fields are each
.B LS_W_NAME
characters wide,
the
.B LS_INUMBER 
field is
.BS LS_W_INUMBER
characters wide,
and the
.B LS_BLOCKS
field is
.B LS_W_BLOCKS
characters wide.
.SH "SEE ALSO"
ls(1), stat(2), strmode(3)


cde/programs/dtksh/ksh93/src/lib/libast/string/fmtls.c
/*
 * CDE - Common Desktop Environment
 *
 * Copyright (c) 1993-2012, The Open Group. All rights reserved.
 *
 * These libraries and programs are free software; you can
 * redistribute them and/or modify them under the terms of the GNU
 * Lesser General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * These libraries and programs are distributed in the hope that
 * they will be useful, but WITHOUT ANY WARRANTY; without even the
 * implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 * PURPOSE. See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with these librararies and programs; if not, write
 * to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
 * Floor, Boston, MA 02110-1301 USA
 */
/* $XConsortium: fmtls.c /main/3 1995/11/01 18:43:43 rswiston $ */
/***************************************************************
*                                                              *
*                      AT&T - PROPRIETARY                      *
*                                                              *
*         THIS IS PROPRIETARY SOURCE CODE LICENSED BY          *
*                          AT&T CORP.                          *
*                                                              *
*                Copyright (c) 1995 AT&T Corp.                 *
*                     All Rights Reserved                      *
*                                                              *
*           This software is licensed by AT&T Corp.            *
*       under the terms and conditions of the license in       *
*       http://www.research.att.com/orgs/ssr/book/reuse        *
*                                                              *
*               This software was created by the               *
*           Software Engineering Research Department           *
*                    AT&T Bell Laboratories                    *
*                                                              *
*               For further information contact                *
*                     gsf@research.att.com                     *
*                                                              *
***************************************************************/

/* : : generated by proto : : */

#if !defined(__PROTO__)
#if defined(__STDC__) || defined(__cplusplus) || defined(_proto) || defined(c_plusplus)
#if defined(__cplusplus)
#define __MANGLE__	"C"
#else
#define __MANGLE__
#endif
#define __STDARG__
#define __PROTO__(x)	x
#define __OTORP__(x)
#define __PARAM__(n,o)	n
#if !defined(__STDC__) && !defined(__cplusplus)
#if !defined(c_plusplus)
#define const
#endif
#define signed
#define void		int
#define volatile
#define __V_		char
#else
#define __V_		void
#endif
#else
#define __PROTO__(x)	()
#define __OTORP__(x)	x
#define __PARAM__(n,o)	o
#define __MANGLE__
#define __V_		char
#define const
#define signed
#define void		int
#define volatile
#endif
#if defined(__cplusplus) || defined(c_plusplus)
#define __VARARG__	...
#else
#define __VARARG__
#endif
#if defined(__STDARG__)
#define __VA_START__(p,a)	va_start(p,a)
#else
#define __VA_START__(p,a)	va_start(p)
#endif
#endif
#include <ast.h>
#include <ls.h>
#include <tm.h>

#ifndef LS_W_MAX
#define LS_W_MAX	128
#endif

/*
 * ls formatter
 *
 *	buf	results placed here
 *	name	file name
 *	st	file stat buffer
 *	info	optional info
 *	link	link text if != 0
 *	flags	LS_* flags
 *
 *	return	end of formatted buf
 */

char*
fmtls __PARAM__((char* buf, const char* name, register struct stat* st, const char* info, const char* link, register int flags), (buf, name, st, info, link, flags)) __OTORP__(char* buf; const char* name; register struct stat* st; const char* info; const char* link; register int flags;){
	register char*		s;
	time_t			tm;

	s = buf;
	if (flags & LS_INUMBER)
		s += sfsprintf(s, LS_W_MAX, "%*lu ", LS_W_INUMBER - 1, (unsigned long)st->st_ino);
	if (flags & LS_BLOCKS)
		s += sfsprintf(s, LS_W_MAX, "%*lu ", LS_W_BLOCKS - 1, (unsigned long)iblocks(st));
	if (flags & LS_LONG)
	{
		s += sfsprintf(s, LS_W_MAX, "%s%3u", fmtmode(st->st_mode, flags & LS_EXTERNAL), (unsigned int)st->st_nlink);
		if (!(flags & LS_NOUSER))
		{
			if (flags & LS_NUMBER) s += sfsprintf(s, LS_W_MAX, " %-*d", LS_W_NAME - 1, st->st_uid);
			else s += sfsprintf(s, LS_W_MAX, " %-*s", LS_W_NAME - 1, fmtuid(st->st_uid));
		}
		if (!(flags & LS_NOGROUP))
		{
			if (flags & LS_NUMBER) s += sfsprintf(s, LS_W_MAX, " %-*d", LS_W_NAME - 1, st->st_gid);
			else s += sfsprintf(s, LS_W_MAX, " %-*s", LS_W_NAME - 1, fmtgid(st->st_gid));
		}
		if (S_ISBLK(st->st_mode) || S_ISCHR(st->st_mode)) s += sfsprintf(s, LS_W_MAX, "%8s ", fmtdev(st));
		else s += sfsprintf(s, LS_W_MAX, "%8lu ", (unsigned long)st->st_size);
		tm = (flags & LS_ATIME) ? st->st_atime : (flags & LS_CTIME) ? st->st_ctime : st->st_mtime;
		s = tmform(s, "%?%l", &tm);
		*s++ = ' ';
	}
	if (info)
	{
		while (*s = *info++) s++;
		*s++ = ' ';
	}
	while (*s = *name++) s++;
	if (flags & LS_MARK)
	{
		if (S_ISDIR(st->st_mode)) *s++ = '/';
#ifdef S_ISLNK
		else if (S_ISLNK(st->st_mode)) *s++ = '@';
#endif
		else if (st->st_mode & (S_IXUSR|S_IXGRP|S_IXOTH)) *s++ = '*';
	}
	if (link)
	{
		s += sfsprintf(s, LS_W_MAX, " %s %s",
#ifdef S_ISLNK
			S_ISLNK(st->st_mode) ? "->" :
#endif
				"==", link);
	}
	*s = 0;
	return(s);
}
